### PR TITLE
SER-1338 | Implement caching methods

### PR DESCRIPTION
**Pull request by slayful**

### CODE REVIEW

File: src/vignette/storage/static_assets.clj [hunk 1]
```diff
@@ -31,10 +51,23 @@
...
+        resource
+        thumb-map
+        mt/thumbnail-path))
+
+  (get-thumbnail [this thumb-map]
+       (get* (:store this)
+         thumb-map
+         mt/thumbnail-path))
+
+  (save-original [this resource original-map]
```
Comment (themech) [inline src/vignette/storage/static_assets.clj]: When does vignette have to save static assets originals? I thought SA caching was needed only for thumbnails.

File: src/vignette/storage/static_assets.clj [hunk 2]
```diff
@@ -31,10 +51,23 @@
...
+  (save-thumbnail [this resource thumb-map]
+      (put* (:store this)
+        resource
+        thumb-map
+        mt/thumbnail-path))
+
+  (get-thumbnail [this thumb-map]
+       (get* (:store this)
+         thumb-map
+         mt/thumbnail-path))
```
Comment (themech) [inline src/vignette/storage/static_assets.clj]: `thumbnail-path` creates a path from keys such as `top-dir` and  `middle-dir`. Do we have these in `thumb-map` for static assets?

File: src/vignette/media_types.clj [hunk 3]
```diff
@@ -152,6 +152,16 @@
         window-width (:window-width data)]
     (window-format width x-offset window-width)))
 
+(defn static-assets-thumb-map->path [data image-path]
+      (let [prefix (thumb-map->prefix data)
+            name (format "%spx-%spx-%s%s"  (or (width data) "X") (or (height data) "X") (mode data) (query-opts-str data))]
```
Comment (slayful) [inline src/vignette/media_types.clj]: so if width or height is not present it's gonna map to `Xpx-Xpx`?

**Comment by slayful** [comment]
LGTM, but don't we want to add a test for caching?

File: src/vignette/storage/s3.clj [hunk 4]
```diff
@@ -105,6 +105,8 @@
   (put-object [this resource bucket path]
     (let [file (file-stream resource)
           mime-type (content-type resource)]
+      (if (perf/timing :s3-get (not (s3/bucket-exists? creds bucket)))
```
Comment (pwojcik86) [inline src/vignette/storage/s3.clj]: I wouldn't time bucket read/creation in the same measurement as file reads/writes

File: src/vignette/storage/static_assets.clj [hunk 5]
```diff
@@ -1,8 +1,33 @@
...
   (:require [vignette.storage.protocols :refer :all]
+            [slingshot.slingshot :refer [throw+]]
             [org.httpkit.client :as http]
             [vignette.util.filesystem :as fs]
-            [clojure.java.io :as io]))
+            [clojure.java.io :as io]
+            [vignette.media-types :as mt]))
+
+(defn get-bucket-name [uuid]
+      (if-let [[id bucket] (re-matches #"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{8}([0-9a-f]{4})" uuid)]
```
Comment (pwojcik86) [inline src/vignette/storage/static_assets.clj]: Use a wildcard `_` instead of `id`, since the value is not used later

**Pull request merged**